transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala

```diff
@@ -15,204 +14,313 @@ import io.mediachain.protocol.Transactor.{JournalError, JournalListener}
+    observers.keys.foreach { observer =>
+      Try(observer.onError(
+        new StatusRuntimeException(
+          Status.UNAVAILABLE.withDescription("Transactor network error")
+        )))
+    }
+    observers = Map()
+    dispatching = Set()
   }
-
-
-  private def publishEvent(event: JournalEvent): Unit = {
-    observers.synchronized {
-      val cancelledObservers: MSet[StreamObserver[JournalEvent]] = MSet()
-      observers.foreach { case (observer, onNext) =>
-        try {
-          onNext(event)
-        } catch {
-          case e: StatusRuntimeException if e.getStatus == Status.CANCELLED =>
-            // if the client killed the connection, remove the observer
-            cancelledObservers.add(observer)
-          case t: Throwable =>
-            throw t
+  
+  private def fetchCurrentBlock() {
+    def setCurrentBlock(block: JournalBlock) {
+      state.index = block.index
+      state.block.clear()
+      state.block ++= block.entries
+      state.blockchain = block.chain
+      state.blockchain.foreach { ref =>
+        emitEvent(TransactorService.journalBlockReferenceToEvent(ref))
+      }
+      state.block.foreach { entry =>
+        emitEvent(TransactorService.journalEntryToEvent(entry))
+      }
+    }
+    
+    def extendCurrentBlock(block: JournalBlock) {
+      val newentries = block.entries.drop(state.block.length)
+      state.index = block.index
+      state.block ++= newentries
+      newentries.foreach { entry =>
+        emitEvent(TransactorService.journalEntryToEvent(entry))
+      }
+    }
+    
+    def fetchBlocks(hd: Option[Reference]) = {
+      def loop(ref: Reference, blocks: List[JournalBlock])
+      : List[JournalBlock]
+      = {
+        val block = getBlock(ref)
+        if (block.chain == state.blockchain) {
+          block :: blocks
+        } else if (block.chain.isEmpty) {
+          throw new RuntimeException("PANIC: Blockchain divergence detected")
+        } else {
+          loop(block.chain.get, block :: blocks)
         }
       }
+      
+      hd match {
+        case Some(ref) => loop(ref, Nil)
+        case None => Nil
+      }
+    }
+    
+    def getBlock(ref: Reference) = {
+      def loop(retry: Int): JournalBlock = {
+        Try(datastore.getAs[JournalBlock](ref)) match {
+          case Success(Some(block)) => block
+          case Success(None) =>
+            val backoff = Client.randomBackoff(retry)
+            logger.info(s"Missing block ${ref}; retrying in ${backoff} ms")
+            Thread.sleep(backoff)
+            loop(retry + 1)
+            
+            // be tolerant of dynamo datastore transient failures
+          case Failure(err: AmazonClientException) =>
+            logger.error("AWS error", err)
+            val backoff = Client.randomBackoff(retry)
+            logger.info(s"Retrying ${ref} in ${backoff} ms")
+            Thread.sleep(backoff)
+            loop(retry + 1)
+            
+          case Failure(err: Throwable) =>
```

Comment:
Don't catch `Throwable`, this will make you catch things like OOM and leave the jvm in an undefined state as your program tries chugging along. Catch `Exception` instead.